### PR TITLE
Fix subtitle sync

### PR DESCRIPTION
**Pull request by redSpoutnik**
Quick PR to fix two problems I've experienced since 10.4.0 upgrade:

* Moving subtitle-sync slider doesn't apply offset to vtt subtitles.
* After closing or fading, subtitle-sync doesn't recover previous subtitle offset.

**Changes**

* Update code to recover subtitle-track.
* The wrong function was tested in "getPlayerSubtitleOffset" in playbackmanager.js (test if actual player support subtitle-sync): it always returned false and subtitleOffset was interpreted as 0.

**Issues**

### CODE REVIEW

File: src/components/htmlvideoplayer/plugin.js [hunk 1]
```diff
@@ -580,18 +580,15 @@ define(['browser', 'require', 'events', 'apphost', 'loading', 'dom', 'playbackMa
...
             .filter(function(trackElement) {
-                if (customTrackIndex === -1 ) {
-                    // get showing .vtt textTacks
-                    return trackElement.mode === 'showing';
-                } else {
-                    // get current .ass textTrack
-                    return ("textTrack" + customTrackIndex) === trackElement.id;
-                }
+                // get showing .vtt textTacks
+                return (trackElement.mode === 'showing') ||
```
Comment (cvium) [inline src/components/htmlvideoplayer/plugin.js]: Redundant parentheses and maybe consider using variables instead of cramming it all in one statement.

**Comment by redSpoutnik** [comment]
I've made changes to fix .ass subtitles sync as well.
I pusshed it here as it is not merged yet, sorry for the multiple commits.
It should be good now.

**Comment by dkanada** [comment]
Is there an easy way to add an option in the user settings to revert to the old behavior? I saw a comment that equated async subtitles to async audio tracks, which obviously isn't often desirable. Sometimes you might want the subtitles to load completely before starting a video to not miss any content.

**Comment by redSpoutnik** [comment]
@dkanada Sorry, but I didn't understand what you meant by revert to old behavior.
Do you mean to disallow the subtitle-sync in player-parameters menu?

**Comment by dkanada** [comment]
My bad, I was thinking of [this](https://github.com/jellyfin/jellyfin-web/pull/463) pull request but that wasn't written by you. Nevermind!

**Comment by redSpoutnik** [comment]
@JustAMan Yes, on Brave 0.69.135 and Chrome 77.0.3865.120.
I've just tested on Firefox 69.0.1 and there is still .vtt sync not working. It seems this is not finished yet...

But maybe it could be merged as a start? I will continue searching anyway.

**Comment by JustAMan** [comment]
If it isn't broken on Firefox (like, it just isn't possible to make an
offset and never was), then it is certainly possible. It just better not be
a regression, that's what I ask.

пн, 14 окт. 2019 г., 19:39 redSpoutnik <notifications@github.com>:

**Comment by dkanada** [comment]
@redSpoutnik are `vtt` subtitles currently working on master or are they broken? If they're broken on master then this isn't a regression and we can merge this for the next release.

File: src/components/htmlvideoplayer/plugin.js [hunk 2]
```diff
@@ -575,36 +575,34 @@ define(['browser', 'require', 'events', 'apphost', 'loading', 'dom', 'playbackMa
...
-                if (customTrackIndex === -1 ) {
+            // if .ass currently rendering
+            if(currentAssRenderer){
+                updateCurrentTrackOffset(offsetValue);
+            } else {
+                var videoElement = self._mediaElement;
+                var mediaStreamTextTracks = getMediaStreamTextTracks(self._currentPlayOptions.mediaSource);
+
+                Array.from(videoElement.textTracks)
+                .filter(function(trackElement) {
```
Comment (dkanada) [inline src/components/htmlvideoplayer/plugin.js]:
```suggestion
                    .filter(function(trackElement) {
```

File: src/components/htmlvideoplayer/plugin.js [hunk 3]
```diff
@@ -575,36 +575,34 @@ define(['browser', 'require', 'events', 'apphost', 'loading', 'dom', 'playbackMa
...
 
             var offsetValue = parseFloat(offset);
-            var videoElement = self._mediaElement;
-            var mediaStreamTextTracks = getMediaStreamTextTracks(self._currentPlayOptions.mediaSource);
 
-            Array.from(videoElement.textTracks)
-            .filter(function(trackElement) {
-                if (customTrackIndex === -1 ) {
+            // if .ass currently rendering
+            if(currentAssRenderer){
```
Comment (dkanada) [inline src/components/htmlvideoplayer/plugin.js]:
```suggestion
            if (currentAssRenderer){
```

File: src/components/htmlvideoplayer/plugin.js [hunk 4]
```diff
@@ -575,36 +575,34 @@ define(['browser', 'require', 'events', 'apphost', 'loading', 'dom', 'playbackMa
...
-                })[0];
+                    var track = customTrackIndex === -1 ? null : mediaStreamTextTracks.filter(function (t) {
+                        return t.Index === customTrackIndex;
+                    })[0];
 
-                if(track) {
-                    offsetValue = updateCurrentTrackOffset(offsetValue);
-                    var format = (track.Codec || '').toLowerCase();
-                    if (format !== 'ass' && format !== 'ssa') {
+                    if(track) {
```
Comment (dkanada) [inline src/components/htmlvideoplayer/plugin.js]:
```suggestion
                    if (track) {
```

**Comment by redSpoutnik** [comment]
@dkanada neither ass nor vtt subtitles are broken with this PR.
.vtt and .ass subtitle sync works fine on chrome and brave, only .ass sync works currently on firefox.

**Comment by JustAMan** [comment]
@redSpoutnik 
Bug does vtt sync work on Firefox _without_ this PR?

**Comment by redSpoutnik** [comment]
@dkanada no it doesn't (broken since 10.4.0).

**Comment by joshuaboniface** [comment]
Cherry-picked into 10.4.1.

**Pull request merged**